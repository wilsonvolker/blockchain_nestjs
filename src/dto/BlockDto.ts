export class BlockDto {
    data: any[];
}