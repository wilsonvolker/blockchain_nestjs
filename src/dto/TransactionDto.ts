export class TransactionDto {
    get amount(): number {
        return this._amount;
    }

    set amount(value: number) {
        this._amount = value;
    }
    get to(): string {
        return this._to;
    }

    set to(value: string) {
        this._to = value;
    }
    get from(): string {
        return this._from;
    }

    set from(value: string) {
        this._from = value;
    }

    private _from: string;
    private _to: string;
    private _amount: number;
    constructor(from: string, to: string, amount: number) {
        this._from = from;
        this._to = to;
        this._amount = amount;
    }
}