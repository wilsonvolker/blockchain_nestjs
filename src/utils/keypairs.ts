// const EC = require("elliptic").ec;
// const ec = new EC("secp256k1");

import * as sha256 from 'crypto-js/sha256';
import {curves, ec} from "elliptic";
const secp256k1: ec  = new ec("secp256k1");

// singleton
export class genSigningKey {
    private static instance: genSigningKey;
    private _signingKeyPair: ec.KeyPair;

    // private _verificationKey: string;
    // private _signingKey: string;

    // get signingKey(): string {
    //     return this._signingKey;
    // }
    //
    // set signingKey(value: string) {
    //     this._signingKey = value;
    // }
    // get verificationKey(): string {
    //     return this._verificationKey;
    // }
    //
    // set verificationKey(value: string) {
    //     this._verificationKey = value;
    // }

    get signingKeyPair(): ec.KeyPair {
        return this._signingKeyPair;
    }

    set signingKeyPair(value: ec.KeyPair) {
        this._signingKeyPair = value;
    }

    public static get Instance(): genSigningKey{
         if (typeof this.instance === "undefined"){
             this.instance = new genSigningKey();
         }
         else{
             return this.instance;
         }
    }

    private constructor() {
        // const keyPair: ec.KeyPair = secp256k1.genKeyPair();
        // this._signingKey = keyPair.getPrivate("hex");
        // this._verificationKey = keyPair.getPublic("hex");
        this._signingKeyPair = secp256k1.genKeyPair();
    }
}

export function sign(keyPair: ec.KeyPair): string {
    return keyPair.sign(sha256(this.from + this.to + this.amount.tostring()), "base64").toDER("hex")
}