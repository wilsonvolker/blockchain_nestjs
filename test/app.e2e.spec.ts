import { Test, TestingModule } from '@nestjs/testing';
import {HttpStatus, INestApplication} from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from './../src/app.module';
import {genSigningKey} from "../src/utils/keypairs.service";
import {TransactionDto} from "../src/dto/TransactionDto";
import {BlockDto} from "../src/dto/BlockDto";
import exp from "constants";

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it('/ (GET)', () => {
    return request(app.getHttpServer())
      .get('/')
      .expect(200)
      .expect('Hello World!');
  });

  it('Test add block of transactions to chain', () => {
    const userKeyPair: genSigningKey = genSigningKey.getKeyPairFromPrivateKey(process.env.INIT_HOLDER_PRIVATE_KEY);
    const dummyKeyPairs: genSigningKey[] = [
      new genSigningKey(),
      new genSigningKey(),
      new genSigningKey(),
    ];


    const tx: TransactionDto[] = [
      new TransactionDto(
          userKeyPair.publicKey,
          dummyKeyPairs[0].publicKey,
          100,
          12
      ),
      new TransactionDto(
          userKeyPair.publicKey,
          dummyKeyPairs[1].publicKey,
          5000,
          0
      ),
      new TransactionDto(
          userKeyPair.publicKey,
          dummyKeyPairs[2].publicKey,
          400
      ),
    ]

    const payload: {block: BlockDto, privateKey: string} = {
      block: {
        data: tx,
      },
      privateKey: userKeyPair.privateKey
    }

    return request(app.getHttpServer())
        .post('/add-block-to-chain')
        .send(payload)
        .expect(HttpStatus.CREATED)
        .then((result) => {
          console.log(result.body)

          const {userBalance, chain} = result.body;
          expect(userBalance).toEqual(100000 - (100 + 12) - 5000 - 400 + (500+12));
          expect(Array.isArray(chain)).toBe(true);
          expect(chain.length > 0).toBe(true);
          chain.forEach((x) => {
            expect(x).toEqual({
              ...x,
              _data: expect.any(Array),
              _hash: expect.any(String),
              _nonce: expect.any(Number),
              // _prevHash: expect.any(String),
              _timestamp: expect.any(String),
            })
          })
        })

        // .expect('Hello World!');
  })

  it("Test check user balance", () => {
    const userKeyPair: genSigningKey = genSigningKey.getKeyPairFromPrivateKey(process.env.INIT_HOLDER_PRIVATE_KEY);

    return request(app.getHttpServer())
        .get(`/get-balance?userPublicKey=${userKeyPair.publicKey}`)
        // .send({
        //   userPublicKey: userKeyPair.publicKey
        // })
        .expect(HttpStatus.OK)
        .then((result) => {
          console.log(result.body)
          const {userPublicKey, balance} = result.body;
          expect(userPublicKey).toEqual(userKeyPair.publicKey);
          expect(balance).toEqual(100000);
        })
  })
});
