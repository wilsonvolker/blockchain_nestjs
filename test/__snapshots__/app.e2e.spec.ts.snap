// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) Test add block of transactions to chain 1`] = `
Object {
  "_data": Any<Array>,
  "_hash": "3329641006d655583dd19bb19158ee673b867e54cfe0fe7105a0b5897996acbc",
  "_nonce": 0,
  "_timestamp": Any<String>,
}
`;

exports[`AppController (e2e) Test add block of transactions to chain 2`] = `
Object {
  "_data": Any<Array>,
  "_hash": "3eded801cbd363f3f80de0a64d5786c4067f0ce144a79f891de37325c143a733",
  "_nonce": 0,
  "_prevHash": "3329641006d655583dd19bb19158ee673b867e54cfe0fe7105a0b5897996acbc",
  "_timestamp": Any<String>,
}
`;
